pkg/ipcache/ipcache.go

```diff
@@ -550,6 +550,36 @@ func (ipc *IPCache) RemoveIdentityOverride(cidr netip.Prefix, identityLabels lab
 	ipc.RemoveMetadata(cidr, resource, overrideIdentity(true), identityLabels)
 }
 
+// AllocateIdentity tries to reserve a given numeric identity for a given prefix. The prefix must already be
+// known to the metadata layer.
+// This takes a reference to the identity. The caller *must* release the allocated identity accordingly.
+// It does not update the actual ipcache; its intended use is to reserve identities during CIDR prefix restoration.
+// If the prefix is unknown or identity allocation has failed, this will return a nil Identity.
+//
+// Calling this function after startup is complete (and thus the prefix has almost certainly had an identity allocated
+// for it) will not be useful; it is only effective before startup when the ipcache is paused.
+func (ipc *IPCache) AllocateIdentity(ctx context.Context, prefix netip.Prefix, existing identity.NumericIdentity) (*identity.Identity, error) {
```

Comment:
@joestringer let me know what you think about this API. The goal is to actually expose the API needed for restoration ("please reserve this cidr -> id mapping") rather than a sketchy mis-use of `AllocateCIDRs` and the reference-counting bugs we've had. If this works out, we can also transition the non-node CIDR restoration logic to use the same API.

Reply:
Note that this relies on the ipcache being stopped (because k8s caches are still syncing). If this is not desirable, we could add a bit of metadata that force-pauses updates for a given prefix. Not perfect, but potentially interesting.

Reply:
I'll continue to mull this over, but I wonder whether longer term the ipcache restore logic should become part of ipcache so that it can (a) dump previous ipcache and understand IP<->NumericIdentity mappings, then (b) pull state from k8s/kvstore, then (c) internally resolve the exact desired labels for each IP + desired NumericIdentity, attempt the allocation of that identity or fall back to another NumericIdentity if necessary.

As a step to allow restore logic and *only* restore logic to inject a "desired NumericIdentity" for a prefix, this seems like a practical way to get it done.

Reply:
I agree that we should consider making restoration part of the ipcache; all of this is tricky since we have to back-fill the labels based on the numeric identity. That is definitely beyond the scope of this change, however.

Reply:
FYI, I added a unit test for this function, making sure it acts as we expect.

Reply:
After a deeper look, and particularly noticing the synchronous identity allocation aspect of this, I'm less convinced that this is the right approach. I _think_ there may be a simpler solution, so I'll take a stab at trying to describe the pitfalls I see and a potential alternative in an overall review.